File: hailmary/src/budget.rs
use crate::error::Result;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct Budget {
    data: BudgetData,
}

#[derive(Serialize, Deserialize)]
struct BudgetData {
    category_groups: Vec<CategoryGroup>,
}

#[derive(Serialize, Deserialize)]
struct CategoryGroup {
    categories: Vec<Category>,
}

#[derive(Serialize, Deserialize)]
struct Category {
    activity: f64,
    category_group_name: String,
    name: String,
}

impl Budget {
    pub fn calculate_spending(&self, ignored_categories: &[String]) -> Result<(f64, f64)> {
        let mut positive_category_sum = 0.0;
        let mut negative_category_sum = 0.0;

        for group in &self.data.category_groups {
            for category in &group.categories {
                if category.activity > 0.0 {
                    if !ignored_categories.contains(&category.category_group_name) {
                        positive_category_sum += category.activity;
                    }
                } else {
                    negative_category_sum += category.activity;
                }
            }
        }

        let actual_inflow = positive_category_sum / 1000.0;
        let actual_outflow = negative_category_sum / 1000.0;

        println!(
            "spending: positive: ${:.2}, negative: ${:.2}",
            actual_inflow, actual_outflow
        );

        let total_spending = actual_inflow + actual_outflow;

        println!("total expense ${:.2}", total_spending);

        Ok((actual_inflow, actual_outflow))
    }

    pub fn from_file(path: &str) -> Result<Budget> {
        let json_data = std::fs::read_to_string(path)?;
        Ok(serde_json::from_str(&json_data)?)
    }
}

File: finny-worker/budget/budget.go
package budget

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"slices"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type BudgetService struct {
	db *gorm.DB
}

func NewBudgetService(db *gorm.DB) *BudgetService {
	return &BudgetService{
		db: db,
	}
}

type Budget struct {
	CategoryGroups []CategoryGroup `json:"category_groups"`
}

type CategoryGroup struct {
	Categories []Category `json:"categories"`
}

type Category struct {
	Activity          float64 `json:"activity"`
	CategoryGroupName string  `json:"category_group_name"`
	Name              string  `json:"name"`
}

type BudgetJson struct {
	Data Budget `json:"data"`
}

type BudgetSQL struct {
	Data Budget `json:"data"`
}

type YNABInflowOutflow struct {
	outflow float64
	inflow  float64
}

func (b *BudgetService) CalculateSpending(budget *Budget, ignoredCategories []string) (YNABInflowOutflow, error) {
	var positiveCategorySum float64
	var negativeCategorySum float64

	for _, group := range budget.CategoryGroups {
		for _, category := range group.Categories {
			if category.Activity > 0.0 {
				if !slices.Contains(ignoredCategories, category.CategoryGroupName) {
					positiveCategorySum += category.Activity
				}
			} else {
				negativeCategorySum += category.Activity
			}
		}
	}

	actualInflow := positiveCategorySum / 1000.0
	actualOutflow := negativeCategorySum / 1000.0

	return YNABInflowOutflow{
		outflow: actualOutflow,
		inflow:  actualInflow,
	}, nil
}

func (b *BudgetService) FromFile(path string) (*Budget, error) {
	jsonData, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var budget BudgetJson
	err = json.Unmarshal(jsonData, &budget)
	if err != nil {
		return nil, err
	}

	return &Budget{
		CategoryGroups: budget.Data.CategoryGroups,
	}, nil
}

func (b *BudgetService) FromDatabase(userID uuid.UUID) (*Budget, error) {
	var result struct {
		CategoriesJSON []byte `gorm:"column:categories_json"`
	}

	err := b.db.Table("ynab_raw").
		Select("categories_json").
		Where("user_id = ?", userID).
		First(&result).Error

	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, fmt.Errorf("no budget found for user ID %s", userID)
		}
		return nil, err
	}

	categoriesJSON := result.CategoriesJSON

	var budgetSQL BudgetSQL
	err = json.Unmarshal(categoriesJSON, &budgetSQL)
	if err != nil {
		return nil, err
	}

	return &Budget{
		CategoryGroups: budgetSQL.Data.CategoryGroups,
	}, nil
}
